pkg/sync/syncWorker.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/xmtp/xmtpd/pkg/registrant"
+	"github.com/xmtp/xmtpd/pkg/registry"
+	"github.com/xmtp/xmtpd/pkg/tracing"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+type syncWorker struct {
+	ctx          context.Context
+	log          *zap.Logger
+	nodeRegistry registry.NodeRegistry
+	registrant   *registrant.Registrant
+	store        *sql.DB
+	wg           sync.WaitGroup
+}
+
+func startSyncWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	nodeRegistry registry.NodeRegistry,
+	registrant *registrant.Registrant,
+	store *sql.DB,
+) (*syncWorker, error) {
+	s := &syncWorker{
+		ctx:          ctx,
+		log:          log.With(zap.String("method", "syncWorker")),
+		nodeRegistry: nodeRegistry,
+		registrant:   registrant,
+		store:        store,
+		wg:           sync.WaitGroup{},
+	}
+	if err := s.start(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *syncWorker) start() error {
+	nodes, err := s.nodeRegistry.GetNodes()
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if node.NodeID == s.registrant.NodeID() || !node.IsHealthy {
+			continue
+		}
+		s.subscribeToNode(node)
+	}
+	return nil
+}
+
+func (s *syncWorker) close() {
+	s.wg.Wait()
+}
+
+func (s *syncWorker) subscribeToNode(node registry.Node) {
+	tracing.GoPanicWrap(
+		s.ctx,
+		&s.wg,
+		fmt.Sprintf("node-subscribe-%d", node.NodeID),
+		func(ctx context.Context) {
+			var err error
+			for {
+				if err != nil {
+					log.Error(fmt.Sprintf("Error: %v, retrying...", err))
+					time.Sleep(1 * time.Second)
+				}
+				err = nil
+
+				conn, err := s.connectToNode(node)
+				if err != nil {
+					continue
+				}
+				stream, err := s.setupStream(node, conn)
+				if err != nil {
+					continue
+				}
+				err = s.listenToStream(stream)
+				if err != nil {
+					continue
+				}
+			}
+		},
+	)
+}
+
+func (s *syncWorker) connectToNode(node registry.Node) (*grpc.ClientConn, error) {
+	addr := node.HttpAddress
+	log.Info(fmt.Sprintf("attempting to connect to %s", addr))
+	conn, err := grpc.NewClient(
+		addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to connect to peer: %v", err)
+	}
+	log.Info(fmt.Sprintf("Successfully connected to peer at %s", addr))
+	return conn, nil
+}
+
+func (s *syncWorker) setupStream(node registry.Node, conn *grpc.ClientConn) (message_api.ReplicationApi_BatchSubscribeEnvelopesClient, error) {
+	client := message_api.NewReplicationApiClient(conn)
+	stream, err := client.BatchSubscribeEnvelopes(
+		s.ctx,
+		&message_api.BatchSubscribeEnvelopesRequest{
+			Requests: []*message_api.BatchSubscribeEnvelopesRequest_SubscribeEnvelopesRequest{
+				{
+					Query: &message_api.EnvelopesQuery{
+						Filter: &message_api.EnvelopesQuery_OriginatorNodeId{
+							OriginatorNodeId: node.NodeID,
+						},
+						LastSeen: nil,
```

Comment:
is this another `todo`?

Reply:
Yes, first we need to have subscriptions that support a cursor, and then implement tracking of lastSeen (done at the same time as validation of originator payloads) - we have work items for these